src/funk/fd_funk_txn.c

```diff
@@ -455,9 +487,105 @@ fd_funk_txn_publish_funk_child( fd_funk_t *     funk,
                                 ulong           tag,
                                 ulong           txn_idx ) {
 
-  /* Publish txn here.  If fail, preserve txn and return an error code. */
+  /* Apply all records (xid,key) in this transaction to (root,key).
+     Notes we don't need to to do all the individual removal pointer
+     updates as we are removing the whole list.  Likewise, we
+     temporarily repurpose txn_cidx as a loop detector for additional
+     corruption protection.  */
+
+  fd_funk_txn_xid_t const * root = fd_funk_root( funk );
+
+  fd_funk_rec_t * rec_map = fd_funk_rec_map( funk, fd_funk_wksp( funk ) ); /* Guarantee we can insert at least 1 as per contract */
+
+  ulong rec_max = funk->rec_max;
+  ulong rec_idx = map[ txn_idx ].rec_head_idx;
+  while( !fd_funk_rec_idx_is_null( rec_idx ) ) {
+
+    /* Validate rec_idx */
+
+    if( FD_UNLIKELY( rec_idx>=rec_max ) ) FD_LOG_CRIT(( "memory corruption detected (bad idx)" ));
+    if( FD_UNLIKELY( fd_funk_txn_idx( rec_map[ rec_idx ].txn_cidx )!=txn_idx ) )
+      FD_LOG_CRIT(( "memory corruption detected (cycle or bad idx)" ));
+    rec_map[ rec_idx ].txn_cidx = fd_funk_txn_cidx( FD_FUNK_TXN_IDX_NULL );
+
+    ulong next_idx = rec_map[ rec_idx ].next_idx;
+
+    /* See if (root,key) already exists */
+
+    fd_funk_xid_key_pair_t pair[1]; fd_funk_xid_key_pair_init( pair, root, fd_funk_rec_key( &rec_map[ rec_idx ] ) );
+    fd_funk_rec_t * root_rec = fd_funk_rec_map_query( rec_map, pair, NULL );
+
+    if( FD_UNLIKELY( rec_map[ rec_idx ].flags & FD_FUNK_REC_FLAG_ERASE ) ) { /* Erase a published key */
+
+      /* Note that we only set the erase flag if there is was ancestor
+         to this transaction with the key in it.  So if we don't find a
+         record, we have a memory corruption problem. */
+
+      if( FD_UNLIKELY( !root_rec ) ) FD_LOG_CRIT(( "memory corruption detected (bad ancestor)" ));
+
+      /* Unmap (xid,key).  Note that value metadata already previously
+         freed when we set erase. */
+
+      fd_funk_rec_map_remove( rec_map, fd_funk_rec_pair( &rec_map[ rec_idx ] ) );
+
+      /* Erase (root,key) */
+
+      /* Free root_rec val resources here */
+
+      ulong prev_idx = root_rec->prev_idx;
+      ulong next_idx = root_rec->next_idx;
 
-  /* At this point, txn has been committed to permanent storage */
+      if( FD_UNLIKELY( fd_funk_rec_idx_is_null( prev_idx ) ) ) funk->rec_head_idx           = next_idx;
```

Comment:
Is rec_head_idx and rec_tail_idx initialized to null before this in this function?

Reply:
They are initialized to NULL at funk creation and they are a list of all the records in the last published transaction at this point.  Which could be NULL or could be something.